dffml/optimizer/parameter_grid.py

```diff
@@ -0,0 +1,94 @@
+from typing import Union, Dict, Any
+import itertools
+import logging
+import sys
+
+from ..base import (
+    config,
+    field,
+)
+from ..high_level.ml import train, accuracy
+from .optimizer import Optimizer, OptimizerConfig, OptimizerContext
+from ..util.entrypoint import entrypoint
+from ..source.source import BaseSource, Record
+from ..accuracy.accuracy import AccuracyScorer, AccuracyContext
+from ..model import ModelContext
+
+
+@config
+class ParameterGridConfig:
+    parameters: dict = field("Parameters to be optimized")
+
+
+class ParameterGridContext(OptimizerContext):
+    """
+    Parameter Grid Optimizer
+    """
+
+    async def optimize(
+        self,
+        model: ModelContext,
+        accuracy_scorer: Union[AccuracyScorer, AccuracyContext],
+        train_data: Union[BaseSource, Record, Dict[str, Any]],
+        test_data: Union[BaseSource, Record, Dict[str, Any]],
+    ):
+        """
+        Method to optimize hyperparameters by parameter grid.
+        Uses a grid of hyperparameters in the form of a dictionary present in config,
+        Trains each permutation of the grid of parameters and compares accuracy.
+        Sets model to the best parameters and returns highest accuracy.
+
+        Parameters
+        ----------
+        model : ModelContext
+            The Model which needs to be used.
+
+        accuracy_scorer: AccuracyContext
+            The accuracy scorer that needs to be used.
+
+        train_data: SourcesContext
+            The train_data to train models on with the hyperparameters provided.
+
+        sources : SourcesContext
+            The test_data to score against and optimize hyperparameters.
+
+        Returns
+        -------
+        float
+            The highest score value
+        """
+        highest_acc = -1
+        best_config = dict()
+        logging.info(
+            f"Optimizing model with parameter grid: {self.parent.config.parameters}"
+        )
+        names = list(self.parent.config.parameters.keys())
+        logging.info(names)
+        with model.config.no_enforce_immutable():
```

Comment:
```suggestion
        # TODO Remove this once we've corrected model configs
        # to make them accurately describe which properties are
        # mutable
        with model.config.no_enforce_immutable():
```

Reply:
I believe this was a miss-click and we want to keep `with model.config.no_enforce_immutable():`